cs/src/Connections/RefreshingTunnelAccessTokenEventArgs.cs

```diff
@@ -0,0 +1,41 @@
+﻿// <copyright file="RefreshingTunnelAccessTokenEventArgs.cs" company="Microsoft">
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// </copyright>
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VsSaaS.TunnelService;
+
+/// <summary>
+/// Event args for tunnel access token refresh event.
+/// </summary>
+public class RefreshingTunnelAccessTokenEventArgs : EventArgs
+{
+    /// <summary>
+    /// Creates a new instance of <see cref="RefreshingTunnelAccessTokenEventArgs"/> class.
+    /// </summary>
+    public RefreshingTunnelAccessTokenEventArgs(string tunnelAccessScope)
+    {
+        TunnelAccessScope = Requires.NotNull(tunnelAccessScope, nameof(tunnelAccessScope));
+    }
+
+    /// <summary>
+    /// Tunnel access scope to get the token for.
+    /// </summary>
+    public string TunnelAccessScope { get; }
+
+    /// <summary>
+    /// Optional token provider function the event handler may set to asynchnronously fetch the token.
+    /// The arguments are: tunnel access scope and cancellation token.
+    /// The result is the task that returns the new tunnel access token or null if it couldn't fetch the token.
+    /// </summary>
+    public Func<string, CancellationToken, Task<string?>>? TunnelAccessTokenProvider { get; set; }
```

Comment:
Is there a reason to make this a `Func<..., Task<string>>` instead of a `Task<string?>`?

For events that can return an async result, I have always used the latter. For example in [SshRequestEventArgs](https://devdiv.visualstudio.com/DevDiv/_git/SSH?path=/src/cs/Ssh/Events/SshRequestEventArgs.cs&version=GBmain&line=64&lineEnd=65&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents)

It does require passing the `CancellationToken` as another property on the event.

Reply:
Oh, this is not what I meant when suggesting a "callback" for the event. For that suggestion there would be a `Func<..., Task<string>>` directly on the `ITunnelHost`. But maybe an event is better.

Reply:
I'll change it to `Task<string?>?`